src/pubsub/mod.rs

```diff
@@ -0,0 +1,337 @@
+        })
+    }
+
+    /// Subscribe to a given topic, returning a Receiver that can be used to consume messages
+    pub fn subscribe(&self, topic: &T) -> Option<Subscriber<M>> {
+        let topic = self.get_topic(topic);
+        // Check if we're at the limit already
+        if topic.subscribers.load(Ordering::SeqCst) >= self.opts.max_subscribers {
+            return None;
+        }
+
+        self.metrics.subscribers.inc();
+        Some(topic.subscribe(self.metrics.subscribers.clone()))
+    }
+
+    /// Tries to send the message to the given topic.
+    /// The publish will fail if there's no subscribers.
+    pub fn publish(&self, topic: &T, message: M) -> Result<usize, SendError<M>> {
+        self.metrics.msgs_sent.inc();
+        self.get_topic(topic).tx.send(message)
+    }
+}
+
+pub struct BrokerBuilder<M, T> {
+    opts: Opts,
+    registry: Registry,
+    _m: PhantomData<M>,
+    _t: PhantomData<T>,
+}
+
+impl<M: Message, T: TopicId> Default for BrokerBuilder<M, T> {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+impl<M: Message, T: TopicId> BrokerBuilder<M, T> {
+    /// Create a new BrokerBuilder
+    pub fn new() -> Self {
+        Self {
+            opts: Opts::default(),
+            registry: Registry::new(),
+            _m: PhantomData,
+            _t: PhantomData,
+        }
+    }
+
+    /// Set the max number of topics supported. Default is 1 million.
+    pub const fn with_max_topics(mut self, max_topics: u64) -> Self {
+        self.opts.max_topics = max_topics;
+        self
+    }
+
+    /// Set the idle timeout when an inactive topic is removed. Default it 10min.
+    pub const fn with_idle_timeout(mut self, idle_timeout: Duration) -> Self {
+        self.opts.idle_timeout = idle_timeout;
+        self
+    }
+
+    /// Set per-topic buffer size. Default is 10k.
+    pub const fn with_buffer_size(mut self, buffer_size: usize) -> Self {
+        self.opts.buffer_size = buffer_size;
+        self
+    }
+
+    /// Set per-topic max subscriber limit. Default is 10k.
+    pub const fn with_max_subscribers(mut self, max_subscribers: usize) -> Self {
+        self.opts.max_subscribers = max_subscribers;
+        self
+    }
+
+    /// Set Prometheus registry to use
+    pub fn with_metric_registry(mut self, registry: &Registry) -> Self {
+        self.registry = registry.clone();
+        self
+    }
+
+    /// Build the Broker
+    pub fn build(self) -> Broker<M, T> {
+        Broker::new(self.opts, &self.registry)
+    }
+}
+
+#[cfg(test)]
+mod test {
+    use super::*;
+
+    #[tokio::test]
+    async fn test_pubsub() {
+        let b: Broker<String, String> = BrokerBuilder::new()
+            .with_buffer_size(3)
+            .with_max_subscribers(1)
+            .build();
+
+        let topic1 = "foo".to_string();
+        let topic2 = "dead".to_string();
+
+        // No subscribers - error
+        assert!(b.publish(&topic1, "".into()).is_err());
```

Comment:
So this is publish only if subscribed model and makes publishers depend on subscribers. What is the rationale, memory efficiency? Why not buffer some messages if subscribers haven't appeared yet?

Reply:
For our use case there's no point in buffering anything. If there are no listeners - then there's no point in sending anything.

We want to send the logs only if there's somebody listening.

Reply:
I've changed the semantics in the latest commit even tighter - the topic will only be created when somebody subscribes to it.

Reply:
> For our use case there's no point in buffering anything.

So historical data (even slightly in the past) won't be available to the audience in this approach, right? Are You sure this is what we want?